streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java

```diff
@@ -140,8 +143,8 @@ boolean hasMoreAvailableCapacityThan(final ClientState other) {
             throw new IllegalStateException("Capacity of other ClientState must be greater than 0");
         }
 
-        final double otherLoad = (double) other.assignedTaskCount() / other.capacity;
-        final double thisLoad = (double) assignedTaskCount() / capacity;
+        final double otherLoad =  ((double) other.assignedTaskCount() + (double) 0.5 * numberOfStateStores) / other.capacity;
```

Comment:
why `0.5` * `numberOfStateStores`? I don't have a problem with this or a suggestion; I just wanted to get your thought process.

Reply:
I thought that state stores might not be treated as a equal to the number of tasks currently assigned to a Client. Therefore, I thought that weighting the importance of the state stores might be more accurate way of judging the load. WDYT? @mjsax  @bbejeck

Reply:
I agree, currently our estimate of the capacity of a client is just its `num.threads` and the workload of all tasks are equal, so the tasks are assigned completely proportional to the num.threads of the client. I think defining the capacity of a client as num.threads are okay, but we need to consider the general problem of "how to estimate the workload of a task", being it active stream task or standby task. And more importantly how to strike the trade-off between workload balance and stickiness.

As a straw-man approach, we can estimate the workload of a stream task based on the number of its state store and the number of its input partitions; the latter is added just to make sure that we do not estimate a stateless task to have zero workload. And for a standby task we can safely estimate the workload purely as the number of its state stores since for stateless tasks its standby will not even be executed. 

Practically I think there is no easy way to compare the workload of a stream task to the workload of a standby task. So I'd suggest we just assign them separately and independently: i.e. when we assign standby tasks we do not consider the assigned active tasks to that client, and if we can get a balanced assignment of active tasks, and a separate and balanced assignment of standby tasks, it is still an close-to-optimal solution.

Following that idea, I'd suggest we 1) assign active tasks and standby tasks completely separately, 2) for active tasks, we can e.g. define its workload as `(num.store + 1) * num.assigned.partitions`, and for standby tasks we can completely define its workload as `num.store`.

Then in this JIRA, we only need to do 1) above, and the standby task part of 2).

WDYT?

Reply:
Oh, @guozhangwang. When you said separately, did you mean that you would want to assign the tasks in such a way that one ClientState will only get ```standByTasks```  while the other receives only active tasks? Or is it that, when calculating the load, we could assign both of them to one ClientState?